File: src/boj/stack/BOJ9012.java
package boj.stack;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class BOJ9012 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            if(isVPS(br))
                System.out.println("YES");
            else
                System.out.println("NO");
        }
    }

    private static boolean isVPS(BufferedReader br) throws IOException {
        Stack<Integer> stack = new Stack<>();
        StringBuilder sb = new StringBuilder(br.readLine());
        for (int i = 0; i < sb.length(); i++) {
            if(sb.charAt(i) == '(')
                stack.push(1);
            else if (sb.charAt(i)==')')
                if(stack.isEmpty())
                    return false;
                else
                    stack.pop();
        }
        return stack.isEmpty();
    }
}

File: ps_kotlin/src/main/kotlin/boj/boj9012.kt
package boj

import java.util.*

fun main() {
    val br = System.`in`.bufferedReader()
    val bw = System.out.bufferedWriter()
    val n = br.readLine().toInt()

    repeat(n) {
        if (isValidParen(br.readLine())) bw.write("YES\n") else bw.write("NO\n")
    }
    bw.flush()
}

fun isValidParen(parens: String): Boolean {
    val st = Stack<Char>()
    for (paren in parens) {
        if (paren == '(') {
            st.push(paren)
        } else if (paren == ')') {
            if (st.isEmpty())
                return false
            st.pop()
        }
    }
    return st.isEmpty()
}
